File: backend/internal/client/pve/node.go
package pve

import (
	"context"
	"fmt"
	"net/url"
)

// GetNodeStatus 获取节点状态
// node: 节点名称
// 返回节点的详细信息（CPU、内存、磁盘、版本等）
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeStatus(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/status", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取节点状态失败: %w", err)
	}
	return result, nil
}

// GetNodeVersion 获取节点 PVE 版本信息
// node: 节点名称
// 返回 PVE 版本号、发行版等信息
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeVersion(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/version", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取节点版本失败: %w", err)
	}
	return result, nil
}

// GetNodeServices 获取节点服务列表
// node: 节点名称
// 返回节点上所有服务的状态信息
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeServices(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/services", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取节点服务列表失败: %w", err)
	}
	return result, nil
}

// GetNodeSyslog 获取节点系统日志
// node: 节点名称, limit: 日志行数限制, start: 起始行号
// 返回系统日志条目列表
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeSyslog(ctx context.Context, node string, limit, start int) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/syslog", node)
	params := url.Values{}
	if limit > 0 {
		params.Set("limit", fmt.Sprintf("%d", limit))
	}
	if start > 0 {
		params.Set("start", fmt.Sprintf("%d", start))
	}
	if err := c.GetWithParams(ctx, path, params, &result); err != nil {
		return nil, fmt.Errorf("获取节点系统日志失败: %w", err)
	}
	return result, nil
}

// GetNodeTasks 获取节点任务列表
// node: 节点名称
// 返回节点上的任务列表
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeTasks(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/tasks", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取节点任务列表失败: %w", err)
	}
	return result, nil
}

// GetNodeTaskStatus 获取指定任务状态
// node: 节点名称, upid: 任务 UPID
// 返回任务的当前状态和退出码
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeTaskStatus(ctx context.Context, node, upid string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/tasks/%s/status", node, upid)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取任务状态失败: %w", err)
	}
	return result, nil
}

// GetNodeTaskLog 获取指定任务日志
// node: 节点名称, upid: 任务 UPID
// 返回任务的日志行列表
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeTaskLog(ctx context.Context, node, upid string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/tasks/%s/log", node, upid)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取任务日志失败: %w", err)
	}
	return result, nil
}

// GetNodeNetwork 获取节点网络接口列表
// node: 节点名称
// 返回所有网络接口的配置信息
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeNetwork(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/network", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取网络接口列表失败: %w", err)
	}
	return result, nil
}

// SetNodeNetworkInterface 配置网络接口
// node: 节点名称, iface: 接口名称, config: 接口配置参数
// 返回异步任务 ID (UPID)
func (c *Client) SetNodeNetworkInterface(ctx context.Context, node, iface string, config NetInterfaceConfig) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/network/%s", node, iface)
	if err := c.Put(ctx, path, config, &upid); err != nil {
		return "", fmt.Errorf("配置网络接口失败: %w", err)
	}
	return upid, nil
}

// CreateNodeNetworkInterface 创建网络接口
// node: 节点名称, iface: 接口名称, config: 接口配置参数
// 返回异步任务 ID (UPID)
func (c *Client) CreateNodeNetworkInterface(ctx context.Context, node, iface string, config NetInterfaceConfig) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/network", node)
	params := make(NetInterfaceConfig)
	params["iface"] = iface
	for k, v := range config {
		params[k] = v
	}
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("创建网络接口失败: %w", err)
	}
	return upid, nil
}

// DeleteNodeNetworkInterface 删除网络接口
// node: 节点名称, iface: 接口名称
// 返回异步任务 ID (UPID)
func (c *Client) DeleteNodeNetworkInterface(ctx context.Context, node, iface string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/network/%s", node, iface)
	if err := c.Delete(ctx, path, &upid); err != nil {
		return "", fmt.Errorf("删除网络接口失败: %w", err)
	}
	return upid, nil
}

// ApplyNodeNetworkChanges 应用网络配置变更
// node: 节点名称
// 返回异步任务 ID (UPID)
func (c *Client) ApplyNodeNetworkChanges(ctx context.Context, node string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/network", node)
	params := map[string]interface{}{"restart": 1}
	if err := c.Put(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("应用网络变更失败: %w", err)
	}
	return upid, nil
}

// GetNodeAPTUpdate 获取可更新的软件包列表
// node: 节点名称
// 返回待更新的软件包列表
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeAPTUpdate(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/apt/update", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取软件包更新列表失败: %w", err)
	}
	return result, nil
}

// NodeUpdatePackages 更新软件包
// node: 节点名称, packages: 要更新的包列表（逗号分隔）
// 返回异步任务 ID (UPID)
func (c *Client) NodeUpdatePackages(ctx context.Context, node string, packages string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/apt/update", node)
	params := map[string]interface{}{"packages": packages}
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("更新软件包失败: %w", err)
	}
	return upid, nil
}

// GetNodeDNS 获取节点 DNS 配置
// node: 节点名称
// 返回 DNS 搜索域和名称服务器配置
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeDNS(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/dns", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取 DNS 配置失败: %w", err)
	}
	return result, nil
}

// SetNodeDNS 更新节点 DNS 配置
// node: 节点名称, config: DNS 配置
// 返回异步任务 ID (UPID)
func (c *Client) SetNodeDNS(ctx context.Context, node string, config DNSConfig) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/dns", node)
	if err := c.Put(ctx, path, config, &upid); err != nil {
		return "", fmt.Errorf("更新 DNS 配置失败: %w", err)
	}
	return upid, nil
}

// GetNodeTime 获取节点时间信息
// node: 节点名称
// 返回节点的时区和时间信息
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeTime(ctx context.Context, node string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/time", node)
	if err := c.Get(ctx, path, &result); err != nil {
		return nil, fmt.Errorf("获取时间信息失败: %w", err)
	}
	return result, nil
}

// SetNodeTimezone 设置节点时区
// node: 节点名称, timezone: 时区（如 Asia/Shanghai）
// 返回异步任务 ID (UPID)
func (c *Client) SetNodeTimezone(ctx context.Context, node, timezone string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/time", node)
	params := map[string]interface{}{"timezone": timezone}
	if err := c.Put(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("设置时区失败: %w", err)
	}
	return upid, nil
}

// GetNodeRRD 获取节点 RRD 性能数据
// node: 节点名称, timeframe: 时间范围 (hour, day, week, month, year)
// dataset: 数据集 (cpu, memory, network, disk, loadavg)
// 使用 interface{} 返回值防止 PVE 9.x 返回意外数据格式时 json.Unmarshal 失败
func (c *Client) GetNodeRRD(ctx context.Context, node string, timeframe, dataset string) (interface{}, error) {
	var result interface{}
	path := fmt.Sprintf("nodes/%s/rrd", node)
	params := url.Values{}
	params.Set("timeframe", timeframe)
	if dataset != "" {
		params.Set("ds", dataset)
	}
	if err := c.GetWithParams(ctx, path, params, &result); err != nil {
		return nil, fmt.Errorf("获取节点 RRD 数据失败: %w", err)
	}
	return result, nil
}

// GetNodeReport 获取节点诊断报告
// node: 节点名称
// 返回节点诊断报告数据
func (c *Client) GetNodeReport(ctx context.Context, node string) ([]byte, error) {
	resp, err := c.Do(ctx, "GET", fmt.Sprintf("nodes/%s/report", node), nil)
	if err != nil {
		return nil, fmt.Errorf("获取节点报告失败: %w", err)
	}
	data, ok := resp.Data.(string)
	if !ok {
		return nil, fmt.Errorf("节点报告数据格式错误")
	}
	return []byte(data), nil
}

// RebootNode 重启节点
// node: 节点名称
// 返回异步任务 ID (UPID)
func (c *Client) RebootNode(ctx context.Context, node string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/reboot", node)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("重启节点失败: %w", err)
	}
	return upid, nil
}

// ShutdownNode 关闭节点
// node: 节点名称
// 返回异步任务 ID (UPID)
func (c *Client) ShutdownNode(ctx context.Context, node string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/shutdown", node)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("关闭节点失败: %w", err)
	}
	return upid, nil
}

File: frontend/src/api/node.ts
/**
 * Node 节点管理 API 端点
 * 提供节点状态、服务、任务、网络、系统配置等管理功能
 */
import { get, post, put } from './request'
import type {
  NodeStatus,
  NodeService,
  NodeSyslog,
  NodeTask,
  TaskLogEntry,
  NetInterface,
  DNSConfig,
  APTUpdate,
  RRDDataPoint,
  QueryOptions,
} from './types'

/**
 * 获取节点状态信息
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getNodeStatus(
  node: string,
  options?: QueryOptions,
): Promise<NodeStatus> {
  return get<NodeStatus>(`/pve/nodes/${node}/status`, undefined, options)
}

/**
 * 获取节点版本信息
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getNodeVersion(
  node: string,
  options?: QueryOptions,
): Promise<{ version: string; release: string; repoid: string }> {
  return get<{ version: string; release: string; repoid: string }>(
    `/pve/nodes/${node}/version`,
    undefined,
    options,
  )
}

/**
 * 获取节点服务列表
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getNodeServices(
  node: string,
  options?: QueryOptions,
): Promise<NodeService[]> {
  return get<NodeService[]>(`/pve/nodes/${node}/services`, undefined, options)
}

/**
 * 获取节点系统日志
 * @param node 节点名称
 * @param params 查询参数 (start, limit, since, until 等)
 * @param options 查询选项
 */
export async function getNodeSyslog(
  node: string,
  params?: { start?: number; limit?: number; since?: number; until?: number; service?: string },
  options?: QueryOptions,
): Promise<NodeSyslog[]> {
  return get<NodeSyslog[]>(`/pve/nodes/${node}/syslog`, params, options)
}

/**
 * 获取节点任务列表
 * @param node 节点名称
 * @param params 查询参数 (start, limit, type, status 等)
 * @param options 查询选项
 */
export async function getNodeTasks(
  node: string,
  params?: { start?: number; limit?: number; type?: string; status?: string },
  options?: QueryOptions,
): Promise<NodeTask[]> {
  return get<NodeTask[]>(`/pve/nodes/${node}/tasks`, params, options)
}

/**
 * 获取任务状态
 * @param node 节点名称
 * @param upid 唯一进程标识符
 * @param options 查询选项
 */
export async function getTaskStatus(
  node: string,
  upid: string,
  options?: QueryOptions,
): Promise<{ upid: string; status: string; exitstatus: string }> {
  return get<{ upid: string; status: string; exitstatus: string }>(
    `/pve/nodes/${node}/tasks/${encodeURIComponent(upid)}/status`,
    undefined,
    options,
  )
}

/**
 * 获取任务日志
 * @param node 节点名称
 * @param upid 唯一进程标识符
 * @param params 查询参数 (start, limit)
 * @param options 查询选项
 */
export async function getTaskLog(
  node: string,
  upid: string,
  params?: { start?: number; limit?: number },
  options?: QueryOptions,
): Promise<TaskLogEntry[]> {
  return get<TaskLogEntry[]>(
    `/pve/nodes/${node}/tasks/${encodeURIComponent(upid)}/log`,
    params,
    options,
  )
}

/**
 * 获取网络接口列表
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getNetworkInterfaces(
  node: string,
  options?: QueryOptions,
): Promise<NetInterface[]> {
  return get<NetInterface[]>(`/pve/nodes/${node}/network`, undefined, options)
}

/**
 * 创建网络接口
 * @param node 节点名称
 * @param params 网络接口参数 (iface, type, address, netmask, gateway 等)
 * @param options 查询选项
 */
export async function createNetworkInterface(
  node: string,
  params: Record<string, unknown>,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/network`, params, options)
}

/**
 * 更新网络接口配置
 * @param node 节点名称
 * @param iface 接口名称
 * @param params 更新参数
 * @param options 查询选项
 */
export async function updateNetworkInterface(
  node: string,
  iface: string,
  params: Record<string, unknown>,
  options?: QueryOptions,
): Promise<string> {
  return put<string>(`/pve/nodes/${node}/network/${iface}`, params, options)
}

/**
 * 删除网络接口
 * @param node 节点名称
 * @param iface 接口名称
 * @param options 查询选项
 */
export async function deleteNetworkInterface(
  node: string,
  iface: string,
  options?: QueryOptions,
): Promise<string> {
  return get<string>(`/pve/nodes/${node}/network/${iface}`, { method: 'delete' } as Record<string, unknown>, options)
}

/**
 * 应用网络配置变更
 * @param node 节点名称
 * @param options 查询选项
 */
export async function applyNetworkChanges(
  node: string,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/network`, { method: 'apply' } as Record<string, unknown>, options)
}

/**
 * 获取 APT 可更新包列表
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getAPTUpdate(
  node: string,
  options?: QueryOptions,
): Promise<APTUpdate[]> {
  return get<APTUpdate[]>(`/pve/nodes/${node}/apt/update`, undefined, options)
}

/**
 * 更新软件包
 * @param node 节点名称
 * @param packages 包名称列表
 * @param options 查询选项
 */
export async function updatePackages(
  node: string,
  packages: string[],
  options?: QueryOptions,
): Promise<string> {
  return post<string>(
    `/pve/nodes/${node}/apt/update`,
    { packages: packages.join(' ') } as Record<string, unknown>,
    options,
  )
}

/**
 * 获取 DNS 配置
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getDNS(
  node: string,
  options?: QueryOptions,
): Promise<DNSConfig> {
  return get<DNSConfig>(`/pve/nodes/${node}/dns`, undefined, options)
}

/**
 * 设置 DNS 配置
 * @param node 节点名称
 * @param config DNS 配置 (dns1, dns2, dns3, search)
 * @param options 查询选项
 */
export async function setDNS(
  node: string,
  config: DNSConfig,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/dns`, config as Record<string, unknown>, options)
}

/**
 * 获取节点时间
 * @param node 节点名称
 * @param options 查询选项
 */
export async function getTime(
  node: string,
  options?: QueryOptions,
): Promise<{ time: number; timezone: string }> {
  return get<{ time: number; timezone: string }>(`/pve/nodes/${node}/time`, undefined, options)
}

/**
 * 操作指定服务（启动/停止/重启）
 * @param node 节点名称
 * @param service 服务名称
 * @param action 操作类型 (start|stop|restart)
 * @param options 查询选项
 */
export async function actionService(
  node: string,
  service: string,
  action: string,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/services/${service}/${action}`, undefined, options)
}

/**
 * 重启指定服务
 * @param node 节点名称
 * @param service 服务名称
 * @param options 查询选项
 */
export async function restartService(
  node: string,
  service: string,
  options?: QueryOptions,
): Promise<string> {
  return actionService(node, service, 'restart', options)
}

/**
 * 启动指定服务
 * @param node 节点名称
 * @param service 服务名称
 * @param options 查询选项
 */
export async function startService(
  node: string,
  service: string,
  options?: QueryOptions,
): Promise<string> {
  return actionService(node, service, 'start', options)
}

/**
 * 停止指定服务
 * @param node 节点名称
 * @param service 服务名称
 * @param options 查询选项
 */
export async function stopService(
  node: string,
  service: string,
  options?: QueryOptions,
): Promise<string> {
  return actionService(node, service, 'stop', options)
}

// ============================================================
// 节点 RRD 监控数据
// ============================================================

/**
 * 获取节点 RRD 性能监控数据
 * @param node 节点名称
 * @param timeframe 时间范围 (hour|day|week|month|year)
 * @param dataset 数据集 (cpu|mem|net|disk|system)
 * @param options 查询选项
 */
export async function getNodeRRD(
  node: string,
  timeframe: string,
  dataset: string,
  options?: QueryOptions,
): Promise<RRDDataPoint[]> {
  return get<RRDDataPoint[]>(
    `/pve/nodes/${node}/rrd`,
    { timeframe, ds: dataset },
    options,
  )
}
